core/src/main/java/ai/timefold/solver/core/impl/score/director/ValueRangeManager.java

```diff
@@ -436,26 +438,41 @@ public ReachableValues getReachableValues(GenuineVariableDescriptor<Solution_> v
                                         entityDescriptor.getEntityClass().getSimpleName(),
                                         variableDescriptor.getVariableName(), valuesSize));
             }
-            var reachableValuesMap = new IdentityHashMap<Object, ReachableItemValue>((int) valuesSize);
+            Class<?> valueClass = null;
+            var idx = 0;
+            while (valueClass == null && idx < allValues.getSize()) {
+                var value = allValues.get(idx++);
+                if (value == null) {
+                    continue;
+                }
+                valueClass = value.getClass();
+                break;
+            }
+            Map<Object, ReachableItemValue> reachableValuesMap = ConfigUtils.isGenericTypeImmutable(valueClass)
+                    ? new HashMap<>((int) valuesSize)
+                    : new IdentityHashMap<>((int) valuesSize);
             for (var entity : entityList) {
```

Comment:
This is a triple-nested loop. The scalability of this for large datasets will cause problems. Long-term, this needs a fix.